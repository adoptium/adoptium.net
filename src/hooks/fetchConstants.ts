import { useEffect, useState } from "react";
import axios from "axios";
import { capitalize } from "@/utils/capitalize";

const baseUrl = "https://api.adoptium.net/v3";

export function useOses(isVisible: boolean): OperatingSystem[] {
  const [oses, setOses] = useState<OperatingSystem[]>([]);

  useEffect(() => {
    let isMounted = true;
    const cancelToken = axios.CancelToken.source();

    if (isVisible) {
      (async () => {
        const url = `${baseUrl}/types/operating_systems`;

        axios
          .get(url, { cancelToken: cancelToken.token })
          .then(function (response) {
            if (!isMounted) return;

            const newOses = response.data
              .map((s: string) => {
                const o: OperatingSystem = {
                  name: capitalize(s),
                  value: s.toLowerCase(),
                };
                return o;
              })
              .sort((o1: OperatingSystem, o2: OperatingSystem) =>
                o1.name.localeCompare(o2.name)
              );

            setOses(newOses);
          })
          .catch(function (error) {
            if (!isMounted || axios.isCancel(error)) return;
            setOses([]);
          });
      })();
    }

    return () => {
      isMounted = false;
      cancelToken.cancel("Component unmounted");
    };
  }, [isVisible]);

  return oses;
}

export function useArches(isVisible: boolean): Architecture[] {
  const [arches, setArches] = useState<Architecture[]>([]);

  useEffect(() => {
    let isMounted = true;
    const cancelToken = axios.CancelToken.source();

    if (isVisible) {
      (async () => {
        const url = `${baseUrl}/types/architectures`;

        axios
          .get(url, { cancelToken: cancelToken.token })
          .then(function (response) {
            if (!isMounted) return;

            const newArches = response.data
              .map((s: string) => {
                const a: Architecture = { name: s, value: s.toLowerCase() };
                if (a.name === "x32") a.name = "x86";
                return a;
              })
              .sort((a1: Architecture, a2: Architecture) =>
                a1.name.localeCompare(a2.name)
              );

            setArches(newArches);
          })
          .catch(function (error) {
            if (!isMounted || axios.isCancel(error)) return;
            setArches([]);
          });
      })();
    }

    return () => {
      isMounted = false;
      cancelToken.cancel("Component unmounted");
    };
  }, [isVisible]);

  return arches;
}

export interface OperatingSystem {
  name: string;
  value: string;
}

export interface Architecture {
  name: string;
  value: string;
}
