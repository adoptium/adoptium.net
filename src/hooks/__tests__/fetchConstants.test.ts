import { renderHook, waitFor } from "@testing-library/react";
import { describe, expect, it, vi, afterAll, afterEach } from "vitest";
import { useArches, useOses } from "../fetchConstants";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";

const mock = new MockAdapter(axios);

afterEach(() => {
  vi.clearAllMocks();
  mock.reset();
});

afterAll(() => {
  mock.restore();
});

describe("useArches", () => {
  it("URL is set correctly", async () => {
    const mockResponse = ["mock_aarch64"];

    mock.onGet().reply(200, mockResponse);
    const spy = vi.spyOn(axios, "get");

    const { result } = renderHook(() => useArches(true));

    await waitFor(
      () => {
        expect(result.current.length).toBeGreaterThan(0);
      },
      { interval: 1 }
    );
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(
      "https://api.adoptium.net/v3/types/architectures"
    );
    expect(result.current).toMatchSnapshot();
  });

  it("Result is empty on error", async () => {
    mock.onGet().reply(500);
    const spy = vi.spyOn(axios, "get");

    const { result } = renderHook(() => useArches(true));
    await waitFor(
      () => {
        expect(result.current.length).toBe(0);
      },
      { interval: 1 }
    );
    expect(spy).toHaveBeenCalledTimes(1);
  });
});

describe("useOses", () => {
  it("URL is set correctly", async () => {
    const mockResponse = ["mock_linux"];

    mock.onGet().reply(200, mockResponse);
    const spy = vi.spyOn(axios, "get");

    const { result } = renderHook(() => useOses(true));

    await waitFor(
      () => {
        expect(result.current.length).toBeGreaterThan(0);
      },
      { interval: 1 }
    );
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(
      "https://api.adoptium.net/v3/types/operating_systems"
    );
    expect(result.current).toMatchSnapshot();
  });

  it("Result is empty on error", async () => {
    mock.onGet().reply(500);
    const spy = vi.spyOn(axios, "get");

    const { result } = renderHook(() => useOses(true));
    await waitFor(
      () => {
        expect(result.current.length).toBe(0);
      },
      { interval: 1 }
    );
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
