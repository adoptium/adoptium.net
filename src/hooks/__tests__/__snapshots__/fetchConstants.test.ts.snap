// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useArches > URL is set correctly 1`] = `
[
  {
    "name": "mock_aarch64",
    "value": "mock_aarch64",
  },
]
`;

exports[`useOses > URL is set correctly 1`] = `
[
  {
    "name": "Mock_linux",
    "value": "mock_linux",
  },
]
`;
